libcontainer/init_linux.go

```diff
@@ -676,3 +688,20 @@ func signalAllProcesses(m cgroups.Manager, s unix.Signal) error {
 
 	return nil
 }
+
+// setupPidfd opens a process file descriptor of init process, and sends the
+// file descriptor back to the socket.
+func setupPidfd(socket *os.File, initType string) error {
+	defer socket.Close()
+
+	pidFd, err := unix.PidfdOpen(os.Getpid(), 0)
+	if err != nil {
+		return fmt.Errorf("failed to pidfd_open: %w", err)
+	}
+
+	if err := utils.SendRawFd(socket, initType, uintptr(pidFd)); err != nil {
```

Comment:
When to close ‘socket’?

Reply:
defer the close at the beginning.

Reply:
Maybe it is safe for `runc create`, `runc exec -d`, and `runc run -d`, but not safe for `runc exec` and `runc run` without detach?
I think it should be closed after we have sent the `pidFd` ASAP.

Reply:
Sorry. I don't follow you here. The defer is used to close the socket. If there is race-condition, the detach should have the same issue. But I didn't see the race-condition here. Would you mind explaining it here? Thanks

Reply:
Because without `detach`, `runc exec` and `runc run` will not exit after the container started.

Reply:
I think you didn't describe the real issue here. I understand the difference between detach and non detach. But the socket has been closed after sending fd. 

The code in the patch is like

1. send pid fd
2. Close pid fd
3. Close socket

The code you want is like

1. send pid fd
2. Close socket 
3. Close pid fd

Yes. I can follow your suggestion.
But I don't see any issue here. And you didn't provide what the issue looks like. If it has, the console socket is the same. Appreciate if you can describe it more instead of describing the difference between detach and non-detach. Thanks

Reply:
You can add a debug log in your local machine, you will find that the defer function is not called after the container started without detach.

Reply:
As we communicated in offline, I see `defer socket.Close()` now, I should pull the latest code to review, it has been closed in 3 days ago.
There is another `defer pidfdSocket.Close()` in `startInitialization`, we have confirmed that `os.File` could be closed twice if there is an error.